File: advertising_sdk/src/main/java/com/adadapted/android/sdk/ui/view/AdZonePresenter.java
package com.adadapted.android.sdk.ui.view;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import com.adadapted.android.sdk.core.ad.Ad;
import com.adadapted.android.sdk.core.ad.AdActionType;
import com.adadapted.android.sdk.core.ad.AdEventClient;
import com.adadapted.android.sdk.core.event.AppEventClient;
import com.adadapted.android.sdk.core.session.Session;
import com.adadapted.android.sdk.core.session.SessionClient;
import com.adadapted.android.sdk.core.session.SessionListener;
import com.adadapted.android.sdk.core.zone.Zone;
import com.adadapted.android.sdk.ui.activity.AaWebViewPopupActivity;
import com.adadapted.android.sdk.ui.messaging.AdContentPublisher;

import java.util.HashMap;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class AdZonePresenter extends SessionListener {
    private static final String LOGTAG = AdZonePresenter.class.getName();

    interface Listener {
        void onZoneAvailable(Zone zone);
        void onAdsRefreshed(Zone zone);
        void onAdAvailable(Ad ad);
        void onNoAdAvailable();
    }

    private String zoneId;
    private final Context context;

    private Listener listener;

    private boolean attached;

    private final Lock zoneLock = new ReentrantLock();
    private String sessionId;
    private boolean zoneLoaded;
    private Zone currentZone;

    private final PixelWebView pixelWebView;

    private int viewCount;

    private Ad currentAd;
    private boolean adStarted;
    private boolean adCompleted;

    private boolean timerRunning;
    private final Lock timerLock = new ReentrantLock();
    private final Timer timer;
    private final AdEventClient adEventClient;
    private final AppEventClient appEventClient;

    AdZonePresenter(final Context context) {
        this.context = context.getApplicationContext();
        adEventClient = AdEventClient.Companion.getInstance();
        appEventClient = AppEventClient.Companion.getInstance();
        pixelWebView = new PixelWebView(context.getApplicationContext());

        attached = false;

        this.zoneLoaded = false;
        this.currentZone = Zone.emptyZone();
        this.viewCount = (int) (Math.random()*10);

        timer = new Timer();
    }

    void init(final String zoneId) {
        if(this.zoneId == null) {
            this.zoneId = zoneId;

            final Map<String, String> params = new HashMap<>();
            params.put("zone_id", zoneId);
            appEventClient.trackSdkEvent("zone_loaded", params);
        }
    }

    void stop() {
        this.onDetach();

        zoneLock.lock();
        try {
            this.zoneId = null;
            this.zoneLoaded = false;
            this.currentZone = Zone.emptyZone();
        }
        finally {
            zoneLock.unlock();
        }
    }

    void onAttach(final Listener l) {
        if(l == null) {
            Log.e(LOGTAG, "NULL Listener provided");
            return;
        }

        zoneLock.lock();
        try {
            if (!attached) {
                attached = true;

                this.listener = l;

                SessionClient.Companion.getInstance().addPresenter(this);
            }

            setNextAd();
        }
        finally {
            zoneLock.unlock();
        }
    }

    void onDetach() {
        zoneLock.lock();
        try {
            if(attached) {
                attached = false;

                this.listener = null;

                completeCurrentAd();
                SessionClient.Companion.getInstance().removePresenter(this);
            }
        }
        finally {
            zoneLock.unlock();
        }
    }

    private void setNextAd() {
        if(!zoneLoaded || timerRunning) {
            return;
        }

        completeCurrentAd();

        zoneLock.lock();
        try {
            if(listener != null && currentZone.hasAds()) {
                final int idx = viewCount % currentZone.getAds().size();
                viewCount++;

                currentAd = currentZone.getAds().get(idx);
            }
            else {
                currentAd = new Ad();
            }

            adStarted = false;
            adCompleted = false;
        }
        finally {
            zoneLock.unlock();
        }

        displayAd();
    }

    private void displayAd() {
        if(currentAd.isEmpty()) {
            notifyNoAdAvailable();
        }
        else {
            notifyAdAvailable(currentAd);
        }
    }

    private void completeCurrentAd() {
        if((currentAd != null && !currentAd.isEmpty()) && (adStarted && !adCompleted)) {
            zoneLock.lock();
            try {
                adCompleted = true;
                AdEventClient.Companion.getInstance().trackImpressionEnd(currentAd);
            }
            finally {
                zoneLock.unlock();
            }
        }
    }

    void onAdDisplayed(final Ad ad) {
        zoneLock.lock();
        try {
            adStarted = true;
            AdEventClient.Companion.getInstance().trackImpression(ad);
            pixelWebView.loadData(ad.getTrackingHtml(), "text/html", null);

            startZoneTimer();
        }
        finally {
            zoneLock.unlock();
        }
    }

    void onAdDisplayFailed(final Ad ad) {
        zoneLock.lock();
        try {
            adStarted = true;
            currentAd = new Ad();

            startZoneTimer();
        }
        finally {
            zoneLock.unlock();
        }
    }

    void onBlankDisplayed() {
        zoneLock.lock();
        try {
            adStarted = true;
            currentAd = new Ad();

            startZoneTimer();
        }
        finally {
            zoneLock.unlock();
        }
    }

    private void startZoneTimer() {
        if(!zoneLoaded || timerRunning) {
            return;
        }

        timerLock.lock();
        try {
            timerRunning = true;
            timer.schedule(new TimerTask(){
                @Override
                public void run() {
                    timerLock.lock();
                    try {
                        timerRunning = false;
                    }
                    finally {
                        timerLock.unlock();
                    }

                    setNextAd();
                }
            }, currentAd.getRefreshTime() * 1000);
        }
        finally {
            timerLock.unlock();
        }
    }

    void onAdClicked(final Ad ad) {
        final String actionType = ad.getActionType();

        final Map<String, String> params = new HashMap<>();
        params.put("ad_id", ad.getId());

        switch(actionType) {
            case AdActionType.CONTENT:
                appEventClient.trackSdkEvent("atl_ad_clicked", params);
                handleContentAction(ad);
                break;

            case AdActionType.LINK:
            case AdActionType.EXTERNAL_LINK:
                adEventClient.trackInteraction(ad);
                handleLinkAction(ad);
                break;

            case AdActionType.POPUP:
                adEventClient.trackInteraction(ad);
                handlePopupAction(ad);
                break;

            case AdActionType.CONTENT_POPUP:
                appEventClient.trackSdkEvent("popup_ad_clicked", params);
                handlePopupAction(ad);
                break;

            default:
                Log.w(LOGTAG, "Cannot handle Action type: " + actionType);
        }
    }

    private void handleContentAction(final Ad ad) {
        final String zoneId = ad.getZoneId();
        AdContentPublisher.getInstance().publishContent(zoneId, ad.getContent());
    }

    private void handleLinkAction(final Ad ad) {
        final Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(ad.getActionPath()));

        context.startActivity(intent);
    }

    private void handlePopupAction(final Ad ad) {
        final Intent intent = AaWebViewPopupActivity.createActivity(context, ad);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        context.startActivity(intent);
    }

    private void notifyZoneAvailable() {
        if(listener != null) {
            listener.onZoneAvailable(currentZone);
        }
    }

    private void notifyAdsRefreshed() {
        if(listener != null) {
            listener.onAdsRefreshed(currentZone);
        }
    }

    private void notifyAdAvailable(final Ad ad) {
        if(listener != null) {
            listener.onAdAvailable(ad);
        }
    }

    private void notifyNoAdAvailable() {
        if(listener != null) {
            listener.onNoAdAvailable();
        }
    }

    private boolean updateSessionId(final String sessionId) {
        zoneLock.lock();
        try {
            if(this.sessionId == null || !this.sessionId.equals(sessionId)) {
                this.sessionId = sessionId;
                return true;
            }
        }
        finally {
            zoneLock.unlock();
        }

        return false;
    }

    private void updateCurrentZone(final Zone zone) {
        zoneLock.lock();
        try {
            this.zoneLoaded = true;
            this.currentZone = zone;
        }
        finally {
            zoneLock.unlock();
        }

        if(currentAd == null || currentAd.isEmpty()) {
            setNextAd();
        }
    }

    @Override
    public void onSessionAvailable(final Session session) {
        updateCurrentZone(session.getZone(zoneId));

        if(updateSessionId(session.getId())) {
            notifyZoneAvailable();
        }
    }

    @Override
    public void onAdsAvailable(final Session session) {
        updateCurrentZone(session.getZone(zoneId));
        notifyAdsRefreshed();
    }

    @Override
    public void onSessionInitFailed() {
        updateCurrentZone(Zone.emptyZone());
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/ui/view/AdZonePresenter.kt
package com.adadapted.android.sdk.ui.view

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import com.adadapted.android.sdk.config.EventStrings
import com.adadapted.android.sdk.core.ad.Ad
import com.adadapted.android.sdk.core.ad.AdActionType
import com.adadapted.android.sdk.core.ad.AdEventClient
import com.adadapted.android.sdk.core.event.AppEventClient
import com.adadapted.android.sdk.core.session.Session
import com.adadapted.android.sdk.core.session.SessionClient
import com.adadapted.android.sdk.core.session.SessionListener
import com.adadapted.android.sdk.core.zone.Zone
import com.adadapted.android.sdk.ui.activity.AaWebViewPopupActivity
import com.adadapted.android.sdk.ui.messaging.AdContentPublisher
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

internal class AdZonePresenter(private val context: Context, private val pixelWebView: PixelWebView = PixelWebView(context.applicationContext), private val aaWebViewPopupActivity: AaWebViewPopupActivity = AaWebViewPopupActivity())
    : SessionListener() {
    internal interface Listener {
        fun onZoneAvailable(zone: Zone)
        fun onAdsRefreshed(zone: Zone)
        fun onAdAvailable(ad: Ad)
        fun onNoAdAvailable()
    }

    private var currentAd: Ad = Ad()
    private var zoneId: String? = null
    private var listener: Listener? = null
    private var attached: Boolean
    private val zoneLock: Lock = ReentrantLock()
    private var sessionId: String? = null
    private var zoneLoaded: Boolean
    private var currentZone: Zone
    private var randomAdStartPosition: Int
    private var adStarted = false
    private var adCompleted = false
    private var timerRunning = false
    private val timerLock: Lock = ReentrantLock()
    private var timer: Timer
    private val adEventClient: AdEventClient = AdEventClient.getInstance()
    private val appEventClient: AppEventClient = AppEventClient.getInstance()
    private val sessionClient: SessionClient = SessionClient.getInstance()

    fun init(zoneId: String) {
        if (this.zoneId == null) {
            this.zoneId = zoneId
            val params: MutableMap<String, String> = HashMap()
            params["zone_id"] = zoneId
            appEventClient.trackSdkEvent(EventStrings.ZONE_LOADED, params)
        }
    }

    fun onAttach(l: Listener?) {
        if (l == null) {
            Log.e(LOGTAG, "NULL Listener provided")
            return
        }
        zoneLock.lock()
        try {
            if (!attached) {
                attached = true
                listener = l
                sessionClient.addPresenter(this)
            }
            setNextAd()
        } finally {
            zoneLock.unlock()
        }
    }

    fun onDetach() {
        zoneLock.lock()
        try {
            if (attached) {
                attached = false
                listener = null
                completeCurrentAd()
                sessionClient.removePresenter(this)
            }
        } finally {
            zoneLock.unlock()
        }
    }

    private fun setNextAd() {
        if (!zoneLoaded || timerRunning) {
            return
        }
        completeCurrentAd()
        zoneLock.lock()
        try {
            currentAd = if (listener != null && currentZone.hasAds()) {
                val adPosition = randomAdStartPosition % currentZone.ads.size
                randomAdStartPosition++
                currentZone.ads[adPosition]
            } else {
                Ad()
            }
            adStarted = false
            adCompleted = false
        } finally {
            zoneLock.unlock()
        }
        displayAd()
    }

    private fun displayAd() {
        if (currentAd.isEmpty) {
            notifyNoAdAvailable()
        } else {
            notifyAdAvailable(currentAd)
        }
    }

    private fun completeCurrentAd() {
        if (!currentAd.isEmpty && adStarted && !adCompleted) {
            zoneLock.lock()
            try {
                if (!currentAd.impressionWasTracked()) {
                    adEventClient.trackInvisibleImpression(currentAd)
                }
                currentAd.resetImpressionTracking() //this is critical to make sure rotating ads can get more than one impression (total)
                adCompleted = true
            } finally {
                zoneLock.unlock()
            }
        }
    }

    fun onAdDisplayed(ad: Ad, isAdVisible: Boolean) {
        zoneLock.lock()
        try {
            adStarted = true
            trackAdImpression(ad, isAdVisible)
            startZoneTimer()
        } finally {
            zoneLock.unlock()
        }
    }

    fun onAdVisibilityChanged(isAdVisible: Boolean) {
        trackAdImpression(currentAd, isAdVisible)
    }

    fun onAdDisplayFailed() {
        zoneLock.lock()
        try {
            adStarted = true
            currentAd = Ad()
            startZoneTimer()
        } finally {
            zoneLock.unlock()
        }
    }

    fun onBlankDisplayed() {
        zoneLock.lock()
        try {
            adStarted = true
            currentAd = Ad()
            startZoneTimer()
        } finally {
            zoneLock.unlock()
        }
    }

    private fun trackAdImpression(ad: Ad, isAdVisible: Boolean) {
        if (!isAdVisible || ad.impressionWasTracked() || ad.isEmpty) return
        ad.setImpressionTracked()
        adEventClient.trackImpression(ad)
        pixelWebView.loadData(ad.trackingHtml, "text/html", null)
    }

    private fun startZoneTimer() {
        if (!zoneLoaded || timerRunning) {
            return
        }
        timerLock.lock()
        try {
            timerRunning = true
            timer.schedule(object : TimerTask() {
                override fun run() {
                    timerLock.lock()
                    timerRunning = try {
                        false
                    } finally {
                        timerLock.unlock()
                    }
                    setNextAd()
                }
            }, currentAd.refreshTime * 1000)
        } finally {
            timerLock.unlock()
        }
    }

    fun onAdClicked(ad: Ad) {
        val actionType = ad.actionType
        val params: MutableMap<String, String> = HashMap()
        params["ad_id"] = ad.id

        when (actionType) {
            AdActionType.CONTENT -> {
                appEventClient.trackSdkEvent(EventStrings.ATL_AD_CLICKED, params)
                handleContentAction(ad)
            }
            AdActionType.LINK, AdActionType.EXTERNAL_LINK -> {
                adEventClient.trackInteraction(ad)
                handleLinkAction(ad)
            }
            AdActionType.POPUP -> {
                adEventClient.trackInteraction(ad)
                handlePopupAction(ad)
            }
            AdActionType.CONTENT_POPUP -> {
                appEventClient.trackSdkEvent(EventStrings.POPUP_AD_CLICKED, params)
                handlePopupAction(ad)
            }
            else -> Log.w(LOGTAG, "Cannot handle Action type: $actionType")
        }

        cycleToNextAdIfPossible()
    }

    private fun cycleToNextAdIfPossible() {
        if (currentZone.ads.count() > 1) {
            restartTimer()
            setNextAd()
        }
    }

    private fun restartTimer() {
        timer.cancel()
        timerRunning = false
        timer = Timer()
    }

    private fun handleContentAction(ad: Ad) {
        val zoneId = ad.zoneId
        AdContentPublisher.getInstance().publishContent(zoneId, ad.getContent())
    }

    private fun handleLinkAction(ad: Ad) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.data = Uri.parse(ad.actionPath)
        context.startActivity(intent)
    }

    private fun handlePopupAction(ad: Ad) {
        val intent = aaWebViewPopupActivity.createActivity(context, ad)
        context.startActivity(intent)
    }

    private fun notifyZoneAvailable() {
        listener?.onZoneAvailable(currentZone)
    }

    private fun notifyAdsRefreshed() {
        listener?.onAdsRefreshed(currentZone)
    }

    private fun notifyAdAvailable(ad: Ad) {
        listener?.onAdAvailable(ad)
    }

    private fun notifyNoAdAvailable() {
        listener?.onNoAdAvailable()
    }

    private fun updateSessionId(sessionId: String): Boolean {
        zoneLock.lock()
        try {
            if (this.sessionId == null || this.sessionId != sessionId) {
                this.sessionId = sessionId
                return true
            }
        } finally {
            zoneLock.unlock()
        }
        return false
    }

    private fun updateCurrentZone(zone: Zone) {
        zoneLock.lock()
        try {
            zoneLoaded = true
            currentZone = zone
        } finally {
            zoneLock.unlock()
        }
        if (currentAd.isEmpty) {
            setNextAd()
        }
    }

    override fun onSessionAvailable(session: Session) {
        updateCurrentZone(session.getZone(zoneId!!))
        if (updateSessionId(session.id)) {
            notifyZoneAvailable()
        }
    }

    override fun onAdsAvailable(session: Session) {
        updateCurrentZone(session.getZone(zoneId!!))
        notifyAdsRefreshed()
    }

    override fun onSessionInitFailed() {
        updateCurrentZone(Zone())
    }

    companion object {
        private val LOGTAG = AdZonePresenter::class.java.name
    }

    init {
        attached = false
        zoneLoaded = false
        currentZone = Zone()
        randomAdStartPosition = (Math.random() * 10).toInt()
        timer = Timer()
    }
}